zerver/lib/actions.py

```diff
@@ -1388,6 +1389,38 @@ def send_user_email_update_event(user_profile: UserProfile) -> None:
     )
 
 
+def send_event_for_updating_delivery_email_fields(
+    user_profile: UserProfile, old_visibility_setting: int, visibility_setting: int
+) -> None:
+    active_users = user_profile.realm.get_active_users()
+    delivery_email_visible_user_ids = []
+    delivery_email_not_visible_user_ids = []
+
+    for user in active_users:
+        if can_access_delivery_email(user, old_visibility_setting) != can_access_delivery_email(
+            user, visibility_setting
+        ):
+            if can_access_delivery_email(user, visibility_setting):
+                delivery_email_visible_user_ids.append(user.id)
+            else:
+                delivery_email_not_visible_user_ids.append(user.id)
+
+    if delivery_email_visible_user_ids != []:
+        person = dict(user_id=user_profile.id, new_delivery_email=user_profile.delivery_email)
+        send_event(
+            user_profile.realm,
+            dict(type="realm_user", op="update", person=person),
+            delivery_email_visible_user_ids,
+        )
+    if delivery_email_not_visible_user_ids != []:
+        person = dict(user_id=user_profile.id, new_delivery_email=None)
+        send_event(
+            user_profile.realm,
+            dict(type="realm_user", op="update", person=person),
+            delivery_email_not_visible_user_ids,
+        )
```

Comment:
Implemented both parts - adding `delivery_email` field if restrictions are relaxed and removing it if restrictions are tightened. Skipping adding part led to some mismatching states in event tests.

Also used `new_delivery_email` parameter here such that it does not interfere with the event that updates `state["delivery_email"]` when user changes their email.

Reply:
Why shouldn't this be identical to that event?  It seems like a client should do the same thing if it receives a new delivery email for another user either way.

Reply:
`state["delivery_email"]` is set to real email even if the user is member and has set their setting to `Admins only`. While for the same user, `delivery_email` of themselves in `raw_users` object is not available.

Reply:
I was actually thinking about this - suppose user is hamlet and is a member. Hamlet sets the setting to `admins only`. Then also the hamlet dict in `state["raw_user"]` available to hamlet does not have `delivery_email` field and due to this in many places like users list in settings, hamlet email will also be hidden to hamlet himself as we do not check separately for the self case and directly use email available with person object. This behavior was present before this PR as well.

Do we want this or should our own email address be visible at all places in UI irrespective of setting?

Reply:
I think the right model is that even with EMAIL_ADDRESS_VISIBILITY_NOBODY, hamlet's own `delivery_email` should be present in his person dictionary?  I'll try to write a bit of a spec in a new comment below.